janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/inmemory/MultiPageEntryBuffer.java

```diff
@@ -0,0 +1,399 @@
+// Copyright 2017 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.diskstorage.keycolumnvalue.inmemory;
+
+import com.google.common.base.Preconditions;
+import org.janusgraph.diskstorage.Entry;
+import org.janusgraph.diskstorage.EntryList;
+import org.janusgraph.diskstorage.StaticBuffer;
+import org.janusgraph.diskstorage.keycolumnvalue.KeySliceQuery;
+
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * "Multi-page" implementation of SharedEntryBuffer.
+ * Maintains a list of pages, each containing a shared byte buffer for maxPageSize Entries. Dividing the buffer into
+ * pages helps maintain performance in the case of very big stores. This is because when mutating, instead of re-copying
+ * all the entries in a single byte buffer, we only locate and modify a few pages, skipping those not affected by the mutation.
+ * <p>
+ * Since only a limited number of stores actually get very big, the parent store switches from SinglePageEntryBuffer
+ * to this implementation when a mutation brings the max number of entries beyond maxPageSize threshold.
+ */
+public class MultiPageEntryBuffer implements SharedEntryBuffer {
+    private static int INITIAL_CAPACITY = 2;
+
+    //NOTE: tracking pages in a list, which internally maintains an array, creates additional overhead;
+    // if we maintain our page list in an array directly, this will save this overhead at the cost of some more boilerplate code here
+    // However in practice the number of multi-page buffers is usually relatively small, so at the moment it seems to be
+    // a reasonable price for having less code
+    private ArrayList<BufferPage> pages;
+
+    public MultiPageEntryBuffer(BufferPage initialPage) {
+        this.pages = new ArrayList<>(INITIAL_CAPACITY);
+        if (initialPage instanceof SinglePageEntryBuffer) {
+            pages.add(new BufferPage(initialPage.getOffsetIndex(), initialPage.getRawData()));
+        } else
+            pages.add(initialPage);
+    }
+
+    MultiPageEntryBuffer(List<BufferPage> allPages) {
+        this.pages = new ArrayList<>(allPages);
+    }
+
+    @Override
+    public int numPages() {
+        return pages.size();
+    }
+
+    @Override
+    public int numEntries() {
+        if (pages == null) {
+            return 0;
+        }
+
+        int numEntries = 0;
+
+        for (BufferPage p : pages) {
```

Comment:
return pages.stream().map(BufferPage::numEntries).collectsum() or something similarshould be possible here.

Reply:
numEntries() is on "critical path" in several places such as in getPageIndex(), and using stream for something as trivial as a sum is probably significantly more overhead than using good old cycle... 
I agree that in a case where we would do this sum once in a while, it would still be arguably worth doing stream just for readability. And I confess that I didn't actually profile to see if there will be any "noticeable(tm)" difference between using stream or not. But still, unless you insist, I would leave this as is. 
Please let me know if you still have concerns